app/models/user_batch.rb

```diff
@@ -35,79 +36,247 @@ def create_users(mission)
+
+          validate_users_batch(users_batch)
+
+          check_uniqueness_on_objects(users_batch, 'email')
+          check_uniqueness_on_objects(users_batch, 'login')
+          check_uniqueness_on_objects(users_batch, 'phone')
+          check_uniqueness_on_objects(users_batch, 'phone2')
+
+          check_uniqueness_on_db(users_batch, 'users', 'email', row_start)
+          check_uniqueness_on_db(users_batch, 'users', 'login', row_start)
+          check_uniqueness_on_db(users_batch, 'users', 'phone', row_start, ['phone','phone2'])
+          check_uniqueness_on_db(users_batch, 'users', 'phone2', row_start, ['phone2','phone'])
+
+          check_validation_errors(users_batch, row_start)
+
+          break if errors_reached_limit
+
+          DirectDBConn.insert_on_db(users_batch, 'users')
+          insert_assignments_on_db(users_batch)
+
+          users.concat users_batch
+
+          p "Batch: #{i}"
+
+          # Set row to start on the next batch
+          row_start = ((i+1) * BATCH_SIZE)+1
+        end
+
+        # now if there was a validation error with any user, rollback the transaction
+        raise ActiveRecord::Rollback if @validation_error
+
+      end # transaction
+    end
+
+    return succeeded?
+  end
+
+  private
+
+  def validate_headers
     # assume the first row is the header row
-    headers = data.row(data.first_row)
+    headers = @data.row(@data.first_row)
 
     expected_headers = Hash[*%i{name phone phone2 email notes}.map do |field|
       [User.human_attribute_name(field), field]
     end.flatten]
 
-    fields = Hash[*headers.map.with_index do |header,index|
+    @fields = Hash[*headers.map.with_index do |header,index|
       [index, expected_headers[header]]
     end.flatten]
 
     # validate headers
-    if fields.values.any?(&:nil?)
+    if @fields.values.any?(&:nil?)
       @validation_error = true
       errors.add(:file, :invalid_headers)
       return succeeded?
     end
+  end
 
-    # create the users in a transaction in case of validation error
-    User.transaction do
-
-      # iterate over each row, creating users as we go
-      data.each_row_streaming(offset: data.first_row).with_index do |row,row_index|
-        # excel row numbers are 1-indexed
-        row_number = 1 + data.first_row + row_index
-
-        # stop processing rows after a certain number of errors
-        # (we do this at the beginning of the loop to avoid adding a redundant
-        # :too_many_errors error on the last loop iteration)
-        if errors.count >= IMPORT_ERROR_CUTOFF
-          # we pass the previous row number to the error message formatting
-          # since the error count cutoff was surpassed by the previous rows,
-          # not this row
-          errors.add(:users, :too_many_errors, row: row_number - 1)
-          break
-        end
+  def number_of_iterations
+    users_rows_count = @data.count - 1
+    (users_rows_count/BATCH_SIZE.to_f).ceil - 1
```

Comment:
Spaces around ops here please. Also this could be one line without loss of readability IMO.